src/providers/yamcs-web-socket.js

```diff
@@ -0,0 +1,115 @@
+const FALLBACK_AND_WAIT_MS = [1000, 5000, 5000, 10000, 10000, 30000];
+
+export default class YamcsWebSocket {
+    constructor(wsUrl) {
+        this._connected = false;
+        this._currentWaitIndex = 0;
+        this._requests = [];
+        this._wsUrl = wsUrl;
+        this._socket;
+
+        this._oncloseSubscribers = new Set();
+        this._onerrorSubscribers = new Set();
+        this._onmessageSubscribers = new Set();
+        this._onopenSubscribers = new Set();
+    }
+
+    createWebsocket() {
+        this._socket = new WebSocket(this._wsUrl);
+
+        this._socket.onopen = () => {
+            clearTimeout(this._reconnectTimeout);
+
+            this._connected = true;
+            console.log(`Established websocket connection to ${this._wsUrl}`);
+
+            this._currentWaitIndex = 0;
+
+            this._onopenSubscribers.forEach(onopenSubscriber => onopenSubscriber.callback());
+        };
+
+        this._socket.onmessage = (event) => {
+            this._onmessageSubscribers.forEach(onmessageSubscriber => onmessageSubscriber.callback(event));
+        };
+
+        this._socket.onerror = (error) => {
+            console.error(error);
+            console.warn("Websocket error, attempting reconnect...");
+            this._connected = false;
+            this._reconnect();
+
+            this.onerrorSubscribers.forEach(onerrorSubscriber => onerrorSubscriber.callback(error));
+        };
+
+        this._socket.onclose = () => {
+            this._connected = false;
+            console.warn("Websocket closed. Attempting to reconnect...");
+            this._reconnect();
+
+            this._oncloseSubscribers.forEach(oncloseSubscriber => oncloseSubscriber.callback());
+        };
+    }
+
+    onclose(callback) {
+        const oncloseCallback = { callback };
+
+        this._oncloseSubscribers.add_(oncloseCallback);
+
+        return () => {
+            this._oncloseSubscribers.delete(oncloseCallback);
+        };
+    }
+
+    onerror(callback) {
+        const onerrorCallback = { callback };
+
+        this.onerrorSubscribers.add(onerrorCallback);
+
+        return () => {
+            this.onerrorSubscribers.delete(onerrorCallback);
+        };
+    }
+
+    onmessage(callback) {
+        const onmessageCallback = { callback };
+
+        this._onmessageSubscribers.add(onmessageCallback);
+
+        return () => {
+            this._onmessageSubscribers.delete(onmessageCallback);
+        };
+    }
+
+    onopen(callback) {
+        const onopenCallback = { callback };
+
+        this._onopenSubscribers.add(onopenCallback);
+
+        return () => {
+            this._onopenSubscribers.delete(onopenCallback);
+        };
+    }
+
+    sendMessage(message) {
```

Comment:
This needs to queue messages if the websocket is not connected right now. The realtime-provider has this logic already, it should be moved here.